Mono.Debugging/Mono.Debugging.Evaluation/AsyncOperationManager.cs

```diff
@@ -27,215 +27,172 @@
+			if (e is OperationCanceledException)
+				return true;
+			var aggregateException = e as AggregateException;
+
+			if (depth > 0 && aggregateException != null) {
+				foreach (var innerException in aggregateException.InnerExceptions) {
+					if (IsOperationCancelledException (innerException, depth - 1))
+						return true;
 				}
 			}
-			else {
-				methodCall.WaitForCompleted (System.Threading.Timeout.Infinite);
+			return false;
+		}
+
+		public OperationResult<TValue> Invoke<TValue> (AsyncOperationBase<TValue> mc, int timeout)
+		{
+			if (timeout <= 0)
+				throw new ArgumentOutOfRangeException("timeout", timeout, "timeout must be greater than 0");
+
+			Task<OperationResult<TValue>> task;
+			var description = mc.Description;
+			var cts = new CancellationTokenSource ();
+			var operationData = new OperationData (mc, cts);
+			lock (currentOperations) {
+				if (disposed)
+					throw new ObjectDisposedException ("Already disposed");
+				DebuggerLoggingService.LogMessage (string.Format("Starting invoke for {0}", description));
+				task = mc.InvokeAsync (cts.Token);
+				currentOperations.Add (operationData);
 			}
 
-			lock (operationsToCancel) {
-				operationsToCancel.Remove (methodCall);
-				ST.Monitor.PulseAll (operationsToCancel);
-				if (methodCall.Aborted) {
+			bool cancelledAfterTimeout = false;
+			try {
+				if (task.Wait (timeout)) {
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} succeeded in {1} ms", description, timeout));
+					return task.Result;
+				}
+				DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} timed out after {1} ms. Cancelling.", description, timeout));
+				cts.Cancel ();
+				try {
+					WaitAfterCancel (mc);
+				}
+				catch (Exception e) {
+					if (IsOperationCancelledException (e)) {
+						DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled after timeout", description));
+						cancelledAfterTimeout = true;
+					}
+					throw;
+				}
+				DebuggerLoggingService.LogMessage (string.Format ("{0} cancelling timed out", description));
+				throw new TimeOutException ();
+			}
+			catch (Exception e) {
+				if (IsOperationCancelledException (e)) {
+					if (cancelledAfterTimeout)
+						throw new TimeOutException ();
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled outside before timeout", description));
 					throw new EvaluatorAbortedException ();
 				}
+				throw;
 			}
-
-			if (!string.IsNullOrEmpty (methodCall.ExceptionMessage)) {
-				throw new Exception (methodCall.ExceptionMessage);
+			finally {
+				lock (currentOperations) {
+					currentOperations.Remove (operationData);
+				}
 			}
 		}
-		
-		public void Dispose ()
+
+
+		public event EventHandler<BusyStateEventArgs> BusyStateChanged = delegate {  };
+
+		void WaitAfterCancel (IAsyncOperationBase op)
 		{
-			Disposing = true;
-			lock (operationsToCancel) {
-				foreach (AsyncOperation op in operationsToCancel) {
-					op.InternalShutdown ();
```

Comment:
Nearly the same logic, but not the same. Let's say an invocation is made. While the caller is waiting for the invocation to end, the manager is disposed, so the call is canceled. But if the call can't really be cancelled, the caller will wait forever.